### ISSUE TITLE
load compressed bytecode

### ISSUE THREAD

**Issue by piranna** [bytenode/bytenode#210]
Add option to compress `.jsc` files with gzip or brotli, and allow to load them. Node.js already has support for compression with the built-in `gzip` module, so we don't need third party dependencies, just only I don't know if detection should be done in the `require()` function and the extensions handler or in the loaders, nor if we should use a special extension nor what one would be correct (`.jsgz`. `.jsbr`, `.jsc.gz`, `.jsc.br`, `.jgz`, `.jcg`, `.bnz`...). What do you think?

**Comment by jjeff** [comment]
Sounds like a nice idea, but I'm curious how useful this would actually be in practice. Text files (such as Javascript) compress quite a bit, but I wonder how binary, and in particular, V8 binary files will compress using gzip. I guess that would be easy enough to test. Have you tried gzipping some .jsc files to see how much space they actually save?

Also, it feels to me like `.jsc.gz` is probably the most semantically-correct suffix. IMHO.

**Comment by OsamaAbbas** [comment]
I agree with Jeff on both points. `jsc.gz` seems the best option for the extension. But, do we really need this feature?

I tried to gzip a bundle file to see how much space could be saved, here is the result:

```
 ls -al bundle.js*
-rw-rw-r-- 1 osama osama 1084296 Mar 28 21:11 bundle.js
-rw-rw-r-- 1 osama osama 1011040 Mar 28 21:11 bundle.jsc
-rw-rw-r-- 1 osama osama  474679 Mar 28 21:11 bundle.jsc.gz
```

So, it's about 1.0 MB as a plain javascript or its compiled version. It's around 474.7 KB when gzipped. Given that bytenode is mainly used in desktop and server-side application, space isn't a real issue there.

**Comment by OsamaAbbas** [comment]
> I don't know if detection should be done in the `require()` function and the extensions handler or in the loaders..

I believe in the `require` function is better.

In the semi-abandoned branch ([2.x](https://github.com/bytenode/bytenode/tree/v2.x)), I have added a feature to [register custom extensions](https://github.com/bytenode/bytenode/blob/v2.x/lib/index.js#L249). This feature can be improved (by hooks for example) to allow doing custom actions (e.g. un-compress the `.gz` file) before running the actual compile bytecode. I wish I could return to update this `2.x` version and release it!

**Comment by piranna** [comment]
> I agree with Jeff on both points. jsc.gz seems the best option for the extension

Ok, let's use `.jsc.gz`/`.jsc.br` 👍🏻 

> So, it's about 1.0 MB as a plain javascript or its compiled version. It's around 474.7 KB when gzipped.

So, bytecode compress REALLY well, like more than a 20:1 ratio, isn't it? Can you do the same with Brotli? It should provide a better compression.

> Given that bytenode is mainly used in desktop and server-side application, space isn't a real issue there.

That's no reason to don't compress it. Also main reason to use bytenode is to don't distribute original source code, so compression can add an extra level of entropy.

(OffTopic: I proposed several years ago that Node.js should be able to use directly the package tarfiles instead of decompress them in the hard disk, both to reduce hard disk usage and improve performance)

> I believe in the require function is better.

Maybe using something like https://www.npmjs.com/package/decompress-maybe (disclaimer, it's mine)?

**Comment by piranna** [comment]
I've though a better aproach: brotly have a better conpression than Gzip, but doesn't have a magic number, that jsc files have. So, no need of new extension: if buffer have the jsc magic number, use It directly, if It doesn't, try to decompress first with brotly. This have the advantage that since there's no magic number, It makes a little bit more difficult to identify and reverse engineer It :-)

**Comment by piranna** [comment]
According to [libMagic](https://github.com/file/file/blob/master/magic/Magdir/javascript), to identify content of a Buffer is v8 bytecode we can check bytes 2 and 3 are `0xC0DE` (little endian). This could be enough, but we can do some steps to extra validate:

1. bytes 0 and 1 are different of zero (external reference table)
2. get start of reservation table/code start and payload size
  1. 40	ulelong&0xFFFFFF00	=0x80000000 -> byte 40 and 28 ulelong
  2. 32	ulelong&0xFFFFFF00	=0x80000000 -> byte 32 and 20 ulelong
  3. 32	ulelong		        =0          -> byte 24 and 16 ulelong
3. add code start and payload size and check is equal to buffer length

Not sure if these extra validation steps would work always or only on the last case due to the reservation tables, luckily it's the used in Node.js 20 and probably all newer versions, but to be safe, we can just only check the `0xC0DE` magic number and to ensure external reference table is different of zero.

**Comment by piranna** [comment]
Implemented at https://github.com/bytenode/bytenode/pull/215. Does we have any plan to add test code coverage metrics?

**Comment by OsamaAbbas** [comment]
I'm in the middle of a trip now and I have an approaching deadline. Give me a couple of days to check and review your work here. I will do my best to come back to you before July (of 2023, hopefully :) ).

**Comment by piranna** [comment]
> before July (of 2023, hopefully :) )

Lol :-P

By the way, in one check I've done, with Brotli I've got to reduce size to 1/3 of initial size, so it's pretty cool :-D

**Comment by piranna** [comment]
Fixed in #215, clossing.
